### DISCUSSION TITLE
Constants for boxed true and false values

### DISCUSSION THREAD

**Discussion by nschuessler** [dotnet/csharplang#4671] [Language Ideas]
We have performance experts going through parts of our code base looking for memory allocations for boxed 'true' and 'false' boolean values and creating constants for the boxed values.

Isn't this something that could be available in the runtime either as constants or as part of the compiler to optimize this so:
1) We can avoid boxing booleans (by using the constants)
2) We can avoid the allocations that happen as a result of boxing.

**Comment by HaloFour** [comment]
This sounds like a suggestion for the https://github.com/dotnet/runtime repository.

To me it sounds like it could be a JIT optimization to detect when a `bool` is being boxed and to push a reference to one of two constant instances on the heap.  However, that would also make boxed `bool`s lose reference identity.

**Comment by Joe4evr** [comment]
This would be ***very*** problematic to do now since [`Unsafe.Unbox<T>()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.unbox?view=net-5.0) can give you a *mutable `ref`* to the heap location.
```cs
using System;
using System.Runtime.CompilerServices;

class Program
{
    private static readonly object BoxedTrue = true;

    static void Main()
    {
        Console.WriteLine(BoxedTrue); //'True'
        Unsafe.Unbox<bool>(BoxedTrue) = false;
        Console.WriteLine(BoxedTrue); //'False'
    }
}
```

See also the discussion at https://github.com/dotnet/runtime/issues/28278

**Reply by yaakov-h** [reply]
Well it is unsafe after all.

**Reply by Joe4evr** [reply]
Sure, and the docs call out that this usage is very discouraged and results in unverifiable IL, but it does work (at the moment) and if the runtime would always make boxed booleans point to the same singleton instances, some malicious API could do something akin to this and break the entire world for that process.

**Reply by jnm2** [reply]
This is already possible with e.g. strings, so how much of an added threat would it really be? If untrusted code is running within the same process, all bets are already off.

**Reply by HaloFour** [reply]
Yep, could always mutate `string.Empty` via reflection.

**Reply by 333fred** [reply]
> Yep, could always mutate `string.Empty` via reflection.

I've been told a good story about how accidentally passing a ref to that through COM (and then subsequent mutation) broke dev10 internal builds. Took quite a bit of sleuthing to find.

**Comment by khellang** [comment]
Duplicate of https://github.com/dotnet/runtime/issues/47596?
